### ISSUE TITLE
How to make ``scatter`` (Just CUDA) results repeatable.

### ISSUE THREAD

**Issue by boliqq07** [rusty1s/pytorch_geometric#2788]
## ❓ Questions & Help

I try to repeat my work, but find  the ``scatter`` in torch_scatter (cuda) is un-stable, though with defined random seed.

Due to the ``scatter``  is in class ``MessagePassing``, I thought It is worth paying attention to.

Or I made mistake or neglected someting?

The following are my test results.

I'd appreciate it, if anyone could help or idea.

[file.md](https://github.com/rusty1s/pytorch_geometric/files/6732282/file.md)

**Comment by rusty1s** [comment]
Scatter is a non-deterministic operation by design since it makes use of atomic operations in which the order of aggregation is non-deterministic, leading to minor numerical differences. As an alternative, you can make use of the `segment_csr` operation of `torch_scatter`, see [here](https://pytorch-scatter.readthedocs.io/en/latest/functions/segment_csr.html).

For message passing layers, deterministic aggregation is only guaranteed when using [`SparseTensor`](https://pytorch-geometric.readthedocs.io/en/latest/notes/sparse_tensor.html).

In the end, I wouldn't worry too much about it. In a deep learning scenario, such numerical instabilities should be only noticeable on really small datasets. Although it is correct that exact reproducible is no longer guaranteed when using non-deterministic operations, we can only force reproducibility for a single permutation (which does not exist in the context of graphs).

**Comment by boliqq07** [comment]
> Scatter is a non-deterministic operation by design since it makes use of atomic operations in which the order of aggregation is non-deterministic, leading to minor numerical differences. As an alternative, you can make use of the `segment_csr` operation of `torch_scatter`, see [here](https://pytorch-scatter.readthedocs.io/en/latest/functions/segment_csr.html).
> 
> For message passing layers, deterministic aggregation is only guaranteed when using [`SparseTensor`](https://pytorch-geometric.readthedocs.io/en/latest/notes/sparse_tensor.html).
> 
> In the end, I wouldn't worry too much about it. In a deep learning scenario, such numerical instabilities should be only noticeable on really small datasets. Although it is correct that exact reproducible is no longer guaranteed when using non-deterministic operations, we can only enforce reproducibility for a single permutation (which does not exist in the context of graphs).

----------------------

```
    def __lift__(self, src, edge_index, dim):
        if isinstance(edge_index, Tensor):
            index = edge_index[dim]
            return src.index_select(self.node_dim, index)
        elif isinstance(edge_index, SparseTensor):
            if dim == 1:
                rowptr = edge_index.storage.rowptr()
                rowptr = expand_left(rowptr, dim=self.node_dim, dims=src.dim())
                return gather_csr(src, rowptr)
            elif dim == 0:
                col = edge_index.storage.col()
                return src.index_select(self.node_dim, col)
        raise ValueError
```

 (This code in ``MessagePassing``)

I try to use ``SparseTensor``, ``segemt_scr`` and ``gather_scr``, I find it indeed deterministic. But I find for one general network, The problem of un-repeatable still exist.

Finally, I find the problem in ``index_select`` function of ``torch``, The ``index_select`` could be the  non-deterministic operation too.

Since I try to use ``[ ]`` but not ``index_select`` and it works though with reduced versatility.

Replace the 
` # return src.index_select(self.node_dim, col)`
by
`return src[col]`

All the thing is OK.

**Comment by rusty1s** [comment]
Running `torch.use_deterministic_algorithms(True)` should fix that as well, I guess :)

**Comment by shuoyinn** [comment]
@rusty1s Hello, you said `scatter` could result in indeterminacy, and thus minor numerical differences occur. But when it comes to `scatter,` intrinsically this operation is permutation invariant as you said, then why there will be difference since it is not depend on the order by which the elements scatter? What I mean, you see, `1 + 2 + 3`, and `2 + 1 + 3`, then gradient and anything will be the same, right? Then why difference occurs?

As far as I know, `cuda` version `scatter` will use several parallel "sub things" to implement this operation, is it the reason?

**Comment by rusty1s** [comment]
Yes, this is due to how floating-point precision works. In case the ordering of operations is not deterministic internally, you may get slightly different outputs, e.g., (1 + 2) + 3 may be different from 1 + (2 + 3).

**Comment by shuoyinn** [comment]
Then I understand, and thank you for your such an immediate reply. Very helpful.

**Comment by jiaqian** [comment]
> > Scatter is a non-deterministic operation by design since it makes use of atomic operations in which the order of aggregation is non-deterministic, leading to minor numerical differences. As an alternative, you can make use of the `segment_csr` operation of `torch_scatter`, see [here](https://pytorch-scatter.readthedocs.io/en/latest/functions/segment_csr.html).
> > For message passing layers, deterministic aggregation is only guaranteed when using [`SparseTensor`](https://pytorch-geometric.readthedocs.io/en/latest/notes/sparse_tensor.html).
> > In the end, I wouldn't worry too much about it. In a deep learning scenario, such numerical instabilities should be only noticeable on really small datasets. Although it is correct that exact reproducible is no longer guaranteed when using non-deterministic operations, we can only enforce reproducibility for a single permutation (which does not exist in the context of graphs).
> 
> ```
>     def __lift__(self, src, edge_index, dim):
>         if isinstance(edge_index, Tensor):
>             index = edge_index[dim]
>             return src.index_select(self.node_dim, index)
>         elif isinstance(edge_index, SparseTensor):
>             if dim == 1:
>                 rowptr = edge_index.storage.rowptr()
>                 rowptr = expand_left(rowptr, dim=self.node_dim, dims=src.dim())
>                 return gather_csr(src, rowptr)
>             elif dim == 0:
>                 col = edge_index.storage.col()
>                 return src.index_select(self.node_dim, col)
>         raise ValueError
> ```
> 
> (This code in `MessagePassing`)
> 
> I try to use `SparseTensor`, `segemt_scr` and `gather_scr`, it is deterministic. But for one general network, The problem of un-repeatable still exist.
> 
> Finally, I find the problem in `index_select` function of `torch`, The `index_select` could be the non-deterministic operation too.
> 
> Since I try to use `[ ]` but not `index_select` and it works though with reduced versatility.
> 
> Replace the ` # return src.index_select(self.node_dim, col)` by `return src[col]`
> 
> All the thing is OK.

Hi, could you please share how to use ` SparseTensor`? I am pretty struggling with it, say some functions like `negative_sampling` only supports tensor rather than ` SparseTensor`.  Where you convert normal tensor to ` SparseTensor`? Thanks:)

**Comment by rusty1s** [comment]
Did you take a look at https://pytorch-geometric.readthedocs.io/en/latest/advanced/sparse_tensor.html? You can convert between the two via:
```python
row, col, edge_attr = adj_t.t().coo()
edge_index = torch.stack([row, col], dim=0)
```

**Comment by jiaqian** [comment]
> Did you take a look at https://pytorch-geometric.readthedocs.io/en/latest/advanced/sparse_tensor.html? You can convert between the two via:
> 
> ```python
> row, col, edge_attr = adj_t.t().coo()
> edge_index = torch.stack([row, col], dim=0)
> ```

Hi Matthias, thanks for the quick reply. I think in my case the big difference between rounds mostly coming from the training data batch generation, do you have any suggestion how to make batch generation (using `NeighborLoader`) deterministic?

**Comment by rusty1s** [comment]
I think `NeighborLoader` should be deterministic if you set a manual seed. Which `pyg-lib`/`torch-sparse` version are you using?

**Comment by rusty1s** [comment]
Deterministic neighborhood sampling is available from `torch-sparse` 0.6.14 onwards, see [here](https://github.com/rusty1s/pytorch_sparse/releases/tag/0.6.14).

**Comment by GregoireLamb** [comment]
> Did you take a look at https://pytorch-geometric.readthedocs.io/en/latest/advanced/sparse_tensor.html? You can convert between the two via:
> 
> ```python
> row, col, edge_attr = adj_t.t().coo()
> edge_index = torch.stack([row, col], dim=0)
> ```


Hello! 

I'm facing a similar issue and wonder if / how to use SparseTensors. What should be transformed to a sparse tensor to get a deterministic output? 

I use `graph = T.Compose([T.ToSparseTensor()])(graph)`  to get an adjacency matrix of type <class 'torch_sparse.tensor.SparseTensor'>.
graph.x (and graph.pos) remain(s) of type <class 'torch.Tensor'> but they are dense and changing them to sparseTensor leads to errors and I found [here ](https://stackoverflow.com/questions/61364160/pytorch-sparse-tensors-do-not-have-strides) that dense tensor should work. 

Idk if it can help but a strange thing is that the behaviors of my different runs only start to diverge after a some iterations as seen in the picture (I checked with a precision of 10^-16 and the 12-15 first loss value are the exact same )

![image](https://github.com/pyg-team/pytorch_geometric/assets/92012843/232960ab-016a-4d5b-939f-31cb95ea643d)

**Comment by rusty1s** [comment]
Yes, `graph.x` should be a dense tensor.

It's hard for me to say what might have gone wrong here. There might be other sources of non-determinism (e.g., differently sampled mini-batches from step 70 onwards).
